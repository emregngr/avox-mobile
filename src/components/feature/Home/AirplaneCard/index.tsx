import React, { memo, useMemo } from 'react'
import { View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import type { TotalAirplane } from '@/types/feature/home'

interface AirplaneCardProps {
  airplane: TotalAirplane
}

export const AirplaneCard = memo(({ airplane }: AirplaneCardProps) => {
  const { count, model } = airplane

  const modelTextProps = useMemo(
    () => ({
      children: model,
      color: 'text-90' as const,
      type: 'h4' as const,
    }),
    [model],
  )

  const countTextProps = useMemo(
    () => ({
      children: count,
      color: 'text-100' as const,
      type: 'h2' as const,
    }),
    [count],
  )

  const cardStyles = useMemo(
    () =>
      'bg-background-primary rounded-xl border border-background-quaternary shadow shadow-background-quaternary flex-row items-center justify-between mb-4 p-4',
    [],
  )

  return (
    <View className={cardStyles}>
      <ThemedText {...modelTextProps} />
      <ThemedText {...countTextProps} />
    </View>
  )
})

AirplaneCard.displayName = 'AirplaneCard'
