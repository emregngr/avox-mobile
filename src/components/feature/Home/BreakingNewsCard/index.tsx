import { Image } from 'expo-image'
import { router } from 'expo-router'
import React, { memo, useCallback, useMemo } from 'react'
import { TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import type { BreakingNews } from '@/types/feature/home'
import { responsive } from '@/utils/common/responsive'

const itemWidth = responsive.deviceWidth

interface BreakingNewsCardProps {
  item: BreakingNews
}

const STATIC_STYLES = {
  cardWidth: { width: itemWidth - 32 },
  containerWidth: { width: itemWidth },
  image: {
    height: '100%' as const,
    width: '100%' as const,
  },
}

export const BreakingNewsCard = memo(({ item }: BreakingNewsCardProps) => {
  const { image, title } = item ?? {}

  const imageSource = useMemo(() => ({ uri: image }), [image])

  const handlePress = useCallback(() => {
    router.navigate({
      params: {
        item: JSON.stringify(item),
      },
      pathname: '/breaking-news-detail',
    })
  }, [item])

  return (
    <View className="items-center px-4" style={STATIC_STYLES.containerWidth}>
      <TouchableOpacity
        activeOpacity={0.7}
        className="h-64 rounded-xl overflow-hidden"
        hitSlop={20}
        onPress={handlePress}
        style={STATIC_STYLES.cardWidth}
      >
        <Image
          cachePolicy="memory-disk"
          contentFit="cover"
          source={imageSource}
          style={STATIC_STYLES.image}
          transition={0}
        />
        <View className="absolute bottom-2 mx-3 bg-background-blur px-2 py-1 overflow-hidden rounded-xl">
          <ThemedText
            color="text-100" ellipsizeMode="tail" numberOfLines={3}
            type="h4"
          >
            {title}
          </ThemedText>
        </View>
      </TouchableOpacity>
    </View>
  )
})

BreakingNewsCard.displayName = 'BreakingNewsCard'
