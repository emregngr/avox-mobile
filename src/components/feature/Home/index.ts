export * from './AirplaneCard'
export * from './BreakingNewsCard'
export * from './DestinationCard'
export * from './HomeAirlineCard'
export * from './HomeAirportCard'
export * from './HomeSlider'
export * from './NewsSection'
export * from './SectionHeader'
export * from './SectionScroll'
