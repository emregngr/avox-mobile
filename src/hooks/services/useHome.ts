import { useQuery } from '@tanstack/react-query'

import { getAllHomeData } from '@/services/homeService'
import useLocaleStore from '@/store/locale'

export function useHome() {
  const { selectedLocale } = useLocaleStore()

  const {
    data: homeData,
    error,
    isLoading,
  } = useQuery({
    queryFn: async () => await getAllHomeData(selectedLocale),
    queryKey: ['homeData', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  return {
    error,
    homeData,
    isLoading,
  }
}
