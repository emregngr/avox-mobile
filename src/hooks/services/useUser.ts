import { getApp } from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  addUserPassword,
  changeUserPassword,
  deleteUserAccount,
  getUser,
  registerDevice,
  registerDeviceToUser,
  updateUser,
} from '@/services/userService'
import { setIsAuthenticated } from '@/store/auth'
import useThemeStore from '@/store/theme'
import { deleteUser } from '@/store/user'
import type { AddPasswordCredentials, ChangePasswordCredentials } from '@/types/feature/password'
import type { ProfileData } from '@/types/feature/user'

const app = getApp()
const auth = getAuth(app)

export const useGetUser = () => {
  const userId = auth?.currentUser?.uid
  return useQuery({
    enabled: !!userId,
    queryFn: getUser,
    queryKey: ['userProfile', userId],
  })
}

export const useUpdateUser = () => {
  const queryClient = useQueryClient()
  const userId = auth?.currentUser?.uid
  const { selectedTheme } = useThemeStore()

  return useMutation<any, Error, ProfileData, { previousProfile: any }>({
    mutationFn: async (profileData: ProfileData) => await updateUser(profileData),
    onError: (error, newProfileData, context) => {
      queryClient.setQueryData(['userProfile', userId], context?.previousProfile)
      Alert.alert(
        getLocale('error'),
        getLocale('profileUpdateFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onMutate: async newProfileData => {
      await queryClient.cancelQueries({ queryKey: ['userProfile', userId] })
      const previousProfile = queryClient.getQueryData(['userProfile', userId])
      queryClient.setQueryData(['userProfile', userId], (oldData: any) => ({
        ...oldData,
        ...newProfileData,
        displayName: `${newProfileData?.firstName} ${newProfileData?.lastName}`,
      }))
      return { previousProfile }
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['userProfile', userId] })
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('profileUpdatedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useChangePassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async (credentials: ChangePasswordCredentials) =>
      await changeUserPassword(credentials),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('passwordChangeFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordChangedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useAddPassword = () => {
  const { selectedTheme } = useThemeStore()

  return useMutation({
    mutationFn: async (credentials: AddPasswordCredentials) => await addUserPassword(credentials),
    onError: () => {
      Alert.alert(
        getLocale('error'),
        getLocale('passwordAddFailed'),
        [
          {
            text: getLocale('ok'),
          },
        ],
        {
          userInterfaceStyle: selectedTheme,
        },
      )
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordAddedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useDeleteUser = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteUserAccount,
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('accountDeleteFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: async () => {
      await deleteUser()
      setIsAuthenticated(false)
      queryClient.clear()
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('accountDeletedSuccessfully'),
        type: 'success',
      })
    },
  })
}

export const useRegisterDevice = () =>
  useMutation({
    mutationFn: async () => await registerDevice(),
  })

export const useRegisterDeviceToUser = () =>
  useMutation({
    mutationFn: async () => await registerDeviceToUser(),
  })
