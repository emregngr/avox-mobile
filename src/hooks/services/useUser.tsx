import firebase from '@react-native-firebase/app'
import { getAuth } from '@react-native-firebase/auth'
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query'
import { router } from 'expo-router'
import { Alert } from 'react-native'
import Toast from 'react-native-toast-message'

import { getLocale } from '@/locales/i18next'
import {
  addUserPassword,
  changeUserPassword,
  deleteUserAccount,
  getUser,
  registerDevice,
  registerDeviceToUser,
  updateUser,
} from '@/services/userService'
import { setIsAuthenticated } from '@/store/auth'
import { deleteUser } from '@/store/user'
import type { AddPasswordCredentials, ChangePasswordCredentials } from '@/types/feature/password'
import type { ProfileData } from '@/types/feature/user'

const auth = getAuth(firebase.app())

export const useGetUser = () => {
  const userId = auth?.currentUser?.uid
  return useQuery({
    enabled: !!userId,
    queryFn: getUser,
    queryKey: ['userProfile', userId],
  })
}

export const useUpdateUser = () => {
  const queryClient = useQueryClient()
  const userId = auth?.currentUser?.uid

  return useMutation<any, Error, ProfileData, { previousProfile: any }>({
    mutationFn: (profileData: ProfileData) => updateUser(profileData),
    onError: (error, newProfileData, context) => {
      queryClient.setQueryData(['userProfile', userId], context?.previousProfile)
      Alert.alert(getLocale('error'), getLocale('profileUpdateFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onMutate: async newProfileData => {
      await queryClient.cancelQueries({ queryKey: ['userProfile', userId] })
      const previousProfile = queryClient.getQueryData(['userProfile', userId])
      queryClient.setQueryData(['userProfile', userId], (oldData: any) => ({
        ...oldData,
        ...newProfileData,
        displayName: `${newProfileData?.firstName} ${newProfileData?.lastName}`,
      }))
      return { previousProfile }
    },
    onSettled: () => {
      queryClient.invalidateQueries({ queryKey: ['userProfile', userId] })
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('profileUpdatedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })
}

export const useChangePassword = () =>
  useMutation({
    mutationFn: (credentials: ChangePasswordCredentials) => changeUserPassword(credentials),
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('passwordChangeFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordChangedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })

export const useAddPassword = () =>
  useMutation({
    mutationFn: (credentials: AddPasswordCredentials) => addUserPassword(credentials),
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('passwordAddFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('passwordAddedSuccessfully'),
        type: 'success',
      })
      router.back()
    },
  })

export const useDeleteUser = () => {
  const queryClient = useQueryClient()

  return useMutation({
    mutationFn: deleteUserAccount,
    onError: () => {
      Alert.alert(getLocale('error'), getLocale('accountDeleteFailed'), [
        {
          text: getLocale('ok'),
        },
      ])
    },
    onSuccess: () => {
      deleteUser()
      setIsAuthenticated(false)
      queryClient.clear()
      Toast.show({
        text1: getLocale('successful'),
        text2: getLocale('accountDeletedSuccessfully'),
        type: 'success',
      })
    },
  })
}

export const useRegisterDevice = () =>
  useMutation({
    mutationFn: registerDevice,
  })

export const useRegisterDeviceToUser = () =>
  useMutation({
    mutationFn: registerDeviceToUser,
  })
