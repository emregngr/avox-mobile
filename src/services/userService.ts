import AsyncStorage from '@react-native-async-storage/async-storage'
import { getApp } from '@react-native-firebase/app'
import {
  EmailAuthProvider,
  getAuth,
  linkWithCredential,
  reauthenticateWithCredential,
  updatePassword,
  updateProfile,
} from '@react-native-firebase/auth'
import {
  deleteDoc,
  doc,
  getDoc,
  getFirestore,
  runTransaction,
  setDoc,
} from '@react-native-firebase/firestore'

import { getLocale } from '@/locales/i18next'
import type { AddPasswordCredentials, ChangePasswordCredentials } from '@/types/feature/password'
import type { ProfileData, UserProfile } from '@/types/feature/user'
import { AnalyticsService } from '@/utils/common/analyticsService'
import Device from '@/utils/common/device'
import { Logger } from '@/utils/common/logger'

const app = getApp()
const auth = getAuth(app)
const db = getFirestore(app)

export const getUser = async (): Promise<UserProfile | null> => {
  const user = auth?.currentUser
  if (!user) {
    return null
  }

  try {
    const userRef = doc(db, 'users', user?.uid)
    const userDoc = await getDoc(userRef)
    await AsyncStorage.setItem('user-id', `${user?.uid}`)

    if (userDoc.exists()) {
      const firestoreData = userDoc.data()
      const userData: UserProfile = {
        createdAt: firestoreData?.createdAt,
        displayName: user.displayName ?? null,
        email: user.email ?? null,
        firstName: firestoreData?.firstName ?? null,
        lastName: firestoreData?.lastName ?? null,
        photoURL: user.photoURL ?? null,
        uid: user.uid,
      }
      await AnalyticsService.setUser(userData)
      return userData
    } else {
      const userData: UserProfile = {
        displayName: user.displayName ?? null,
        email: user.email ?? null,
        firstName: null,
        lastName: null,
        photoURL: user.photoURL ?? null,
        uid: user.uid,
      }
      await AnalyticsService.setUser(userData)
      return userData
    }
  } catch (error) {
    Logger.breadcrumb('Failed to get user', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const updateUser = async (profileData: ProfileData): Promise<void> => {
  const user = auth?.currentUser
  if (!user) {
    throw new Error(getLocale('userNotLoggedIn'))
  }

  const { email, firstName, lastName } = profileData
  const userRef = doc(db, 'users', user?.uid)

  try {
    await Promise.all([
      setDoc(userRef, { email, firstName, lastName }, { merge: true }),
      updateProfile(user, { displayName: `${firstName} ${lastName}` }),
    ])
  } catch (error) {
    Logger.breadcrumb('Failed to update user', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const changeUserPassword = async (credentials: ChangePasswordCredentials): Promise<void> => {
  const { currentPassword, newPassword } = credentials
  const user = auth?.currentUser

  if (!user?.email) {
    throw new Error(getLocale('userNotLoggedIn'))
  }

  try {
    const credential = EmailAuthProvider.credential(user.email, currentPassword)
    await reauthenticateWithCredential(user, credential)
    await updatePassword(user, newPassword)
  } catch (error) {
    Logger.breadcrumb('Failed to change user password', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const addUserPassword = async (credentials: AddPasswordCredentials): Promise<void> => {
  const { newPassword } = credentials
  const user = auth?.currentUser

  if (!user?.email) {
    throw new Error(getLocale('userNotLoggedIn'))
  }

  try {
    const credential = EmailAuthProvider.credential(user?.email, newPassword)
    await linkWithCredential(user, credential)
  } catch (error) {
    Logger.breadcrumb('Failed to add user password', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const deleteUserAccount = async (): Promise<void> => {
  const user = auth?.currentUser
  if (!user) {
    throw new Error(getLocale('userNotLoggedIn'))
  }

  try {
    const userDocRef = doc(db, 'users', user?.uid)
    await deleteDoc(userDocRef)
    await user.delete()
  } catch (error) {
    Logger.breadcrumb('Failed to delete user account', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const registerDevice = async (): Promise<void> => {
  try {
    const deviceParams = await Device.registerDevice()
    const deviceId = deviceParams?.unique_id

    if (!deviceId) {
      throw new Error(getLocale('somethingWentWrong'))
    }

    const deviceRef = doc(db, 'devices', deviceId)

    const deviceData = {
      ...deviceParams,
      createdAt: new Date(),
      updatedAt: new Date(),
    }

    await setDoc(deviceRef, deviceData)
  } catch (error) {
    Logger.breadcrumb('Failed to register device', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}

export const registerDeviceToUser = async (): Promise<void> => {
  const user = auth.currentUser
  if (!user) {
    throw new Error(getLocale('userNotLoggedIn'))
  }

  try {
    const deviceParams = await Device.registerDevice()
    const deviceId = deviceParams?.unique_id

    if (!deviceId) {
      throw new Error(getLocale('somethingWentWrong'))
    }

    const userRef = doc(db, 'users', user?.uid)
    const deviceRef = doc(db, 'devices', deviceId)

    const updatedDevice = {
      ...deviceParams,
      updatedAt: new Date(),
    }

    await runTransaction(db, async transaction => {
      const userDoc = await transaction.get(userRef)
      const devices = userDoc?.data()?.devices ?? []

      const existingDeviceIndex = devices?.findIndex(
        (device: typeof updatedDevice) => device?.unique_id === deviceId,
      )

      if (existingDeviceIndex > -1) {
        devices[existingDeviceIndex] = updatedDevice
      } else {
        devices.push(updatedDevice)
      }

      transaction.set(userRef, { devices }, { merge: true })

      transaction.update(deviceRef, {
        updatedAt: new Date(),
        userId: user.uid,
      })
    })
  } catch (error) {
    Logger.breadcrumb('Failed to register device to user', 'error', error as Error)
    throw new Error(getLocale('somethingWentWrong'))
  }
}
