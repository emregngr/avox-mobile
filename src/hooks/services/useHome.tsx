import { useQuery } from '@tanstack/react-query'

import { getAllHomeData } from '@/services/homeService'
import useLocaleStore from '@/store/locale'

export function useHome() {
  const { selectedLocale } = useLocaleStore()

  const {
    data: homeData,
    error,
    isLoading,
  } = useQuery({
    queryFn: () => getAllHomeData(selectedLocale),
    queryKey: ['home', selectedLocale],
    staleTime: 5 * 60 * 1000,
  })

  return {
    error,
    homeData,
    isLoading,
  }
}
