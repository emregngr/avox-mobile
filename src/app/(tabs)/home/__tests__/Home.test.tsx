import { fireEvent, render, waitFor } from '@testing-library/react-native'
import { router } from 'expo-router'
import type { ReactNode } from 'react'
import React from 'react'
import { SafeAreaProvider } from 'react-native-safe-area-context'

import Home from '@/app/(tabs)/home'
import { useHome } from '@/hooks/services/useHome'
import { useRegisterDevice, useRegisterDeviceToUser } from '@/hooks/services/useUser'
import { registerDevice, registerDeviceToUser } from '@/services/userService'
import useAuthStore from '@/store/auth'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('@/store/auth')

const mockedUseAuthStore = useAuthStore as jest.MockedFunction<typeof useAuthStore>

jest.mock('@/hooks/services/useHome')

const mockedUseHome = useHome as jest.MockedFunction<typeof useHome>

jest.mock('@/hooks/services/useUser')

const mockedUseRegisterDevice = useRegisterDevice as jest.MockedFunction<typeof useRegisterDevice>

const mockedUseRegisterDeviceToUser = useRegisterDeviceToUser as jest.MockedFunction<
  typeof useRegisterDeviceToUser
>

jest.mock('@/services/userService')

const mockedRegisterDevice = registerDevice as jest.MockedFunction<typeof registerDevice>

const mockedRegisterDeviceToUser = registerDeviceToUser as jest.MockedFunction<
  typeof registerDeviceToUser
>

jest.mock('@/components/common', () => {
  const { View } = require('react-native')

  return {
    FullScreenLoading: (props: any) => <View {...props} testID="full-screen-loading" />,
  }
})

jest.mock('@/components/feature', () => {
  const { View, Text, TouchableOpacity } = require('react-native')

  return {
    NewsSection: ({ breakingNews }: { breakingNews: any[] }) => (
      <View testID="news-section">
        <Text>News Count: {breakingNews.length}</Text>
      </View>
    ),

    SectionScroll: ({
      title,
      onViewAll,
      data,
    }: {
      title: string
      onViewAll: () => void
      data: any[]
    }) => (
      <View testID={`section-${title}`}>
        <Text>{title}</Text>
        <TouchableOpacity onPress={onViewAll} testID={`view-all-${title}`}>
          <Text>View All</Text>
        </TouchableOpacity>
        <Text>Item count: {data.length}</Text>
      </View>
    ),
  }
})

const mockedHomeData = {
  breakingNews: [{ id: 1, title: 'Test News' }],
  popularAirlines: [{ id: 'TK', name: 'Turkish Airlines' }],
  popularAirports: [{ id: 'IST', name: 'Istanbul Airport' }],
  popularDestinations: [{ id: 'AYT', name: 'Antalya' }],
  totalAirplanes: [{ id: 'TC-JJA', model: 'A330' }],
}

const renderWithSafeAreaProvider = (component: ReactNode) =>
  render(
    <SafeAreaProvider
      initialMetrics={{
        insets: { top: 0, left: 0, right: 0, bottom: 0 },
        frame: { x: 0, y: 0, width: 0, height: 0 },
      }}
    >
      {component}
    </SafeAreaProvider>,
  )

beforeEach(() => {
  mockedUseHome.mockReturnValue({
    homeData: null,
    isLoading: true,
  } as any)

  mockedUseRegisterDevice.mockReturnValue({ mutateAsync: mockedRegisterDevice } as any)

  mockedUseRegisterDeviceToUser.mockReturnValue({ mutateAsync: mockedRegisterDeviceToUser } as any)

  mockedUseAuthStore.mockReturnValue({ isAuthenticated: false })
})

describe('Home Screen', () => {
  it('should display FullScreenLoading component while data is loading', () => {
    const { getByTestId } = renderWithSafeAreaProvider(<Home />)
    expect(getByTestId('full-screen-loading')).toBeTruthy()
  })

  it('should display all sections with correct titles when data is loaded', () => {
    mockedUseHome.mockReturnValue({ homeData: mockedHomeData, isLoading: false } as any)

    const { getByText, getByTestId } = renderWithSafeAreaProvider(<Home />)

    expect(getByTestId('news-section')).toBeTruthy()
    expect(getByText('popularAirlines')).toBeTruthy()
    expect(getByText('popularAirports')).toBeTruthy()
    expect(getByText('popularDestinations')).toBeTruthy()
    expect(getByText('totalAirplanes')).toBeTruthy()
  })

  it('should call only handleAddDevice when the user is not authenticated', async () => {
    mockedUseHome.mockReturnValue({ homeData: mockedHomeData, isLoading: false } as any)
    mockedUseAuthStore.mockReturnValue({ isAuthenticated: false })

    renderWithSafeAreaProvider(<Home />)

    await waitFor(() => {
      expect(mockedRegisterDevice).toHaveBeenCalledTimes(1)
      expect(mockedRegisterDeviceToUser).not.toHaveBeenCalled()
    })
  })

  it('should call both handleAddDevice and handleRegisterDeviceToUser when the user is authenticated', async () => {
    mockedUseHome.mockReturnValue({ homeData: mockedHomeData, isLoading: false } as any)
    mockedUseAuthStore.mockReturnValue({ isAuthenticated: true })

    renderWithSafeAreaProvider(<Home />)

    await waitFor(() => {
      expect(mockedRegisterDevice).toHaveBeenCalledTimes(1)
      expect(mockedRegisterDeviceToUser).toHaveBeenCalledTimes(1)
    })
  })

  it('should navigate to the correct pages when "View All" buttons are pressed', () => {
    mockedUseHome.mockReturnValue({ homeData: mockedHomeData, isLoading: false } as any)

    const { getByTestId } = renderWithSafeAreaProvider(<Home />)

    fireEvent.press(getByTestId('view-all-popularAirlines'))
    expect(router.navigate).toHaveBeenCalledWith('/all-popular-airlines')

    fireEvent.press(getByTestId('view-all-popularAirports'))
    expect(router.navigate).toHaveBeenCalledWith('/all-popular-airports')

    fireEvent.press(getByTestId('view-all-popularDestinations'))
    expect(router.navigate).toHaveBeenCalledWith('/all-popular-destinations')

    fireEvent.press(getByTestId('view-all-totalAirplanes'))
    expect(router.navigate).toHaveBeenCalledWith('/total-airplanes')

    expect(router.navigate).toHaveBeenCalledTimes(4)
  })
})

describe('Home Screen Snapshot', () => {
  it('should render the Home screen successfully', () => {
    const { toJSON } = renderWithSafeAreaProvider(<Home />)

    expect(toJSON()).toMatchSnapshot()
  })
})
