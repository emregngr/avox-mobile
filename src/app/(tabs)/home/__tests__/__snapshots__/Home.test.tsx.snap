// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home Screen Snapshot should render the Home screen successfully 1`] = `
<View
  testID="full-screen-loading"
/>
`;
